src/snapshot.js

```diff
@@ -1,57 +1,65 @@
 'use strict'
 
 const { Session } = require('inspector')
-const { openSync, closeSync, writeSync } = require('fs')
+const SonicBoom = require('sonic-boom')
 const { ensurePromiseCallback, destinationFile } = require('./utils')
 
-module.exports = function generateHeapSnapshot(cb) {
+module.exports = function generateHeapSnapshot(options, cb) {
+  /* istanbul ignore if */
+  if (typeof options === 'function') {
+    cb = options
+    options = null
+  }
+
   // Prepare the context
   const [callback, promise] = ensurePromiseCallback(cb)
-  const destination = destinationFile('heapsnapshot')
+  const { destination } = Object.assign({ destination: destinationFile('heapsnapshot') }, options)
   const session = new Session()
-  let fd
-  let writeError
-
-  // Open the destination file
-  try {
-    fd = openSync(destination, 'w')
-  } catch (e) {
-    return callback(e)
+  let error = null
+
+  if (typeof destination !== 'string' || destination.length === 0) {
+    throw new Error('The destination option must be a non empty string')
   }
 
+  const writer = new SonicBoom({ dest: destination })
+
+  writer.on('error', err => {
+    callback(err)
+  })
+
+  writer.on('close', () => {
+    callback(error, destination)
+  })
```

Comment:
you need to remove those listeners to handle the case of 'close' and 'error' are both emitted.

Reply:
Handle in a equivalent way.